### ISSUE TITLE
[Marlin] Can't decrypt data

### ISSUE THREAD

**Issue by TheGabrielHoward** [TeamWin/Team-Win-Recovery-Project#1476]
- [x] I am running an official build of TWRP, downloaded from https://twrp.me/Devices/
- [x] I am running the latest version of TWRP
- [x] I have read the FAQ (https://twrp.me/FAQ/)
- [x] I have searched for my issue and it does not already exist

Marlin - Pixel XL
3.3.0-0

#### WHAT STEPS WILL REPRODUCE THE PROBLEM?
- With a pin/pattern go to twrp
- Put your pin/pattern
- See the error

#### WHAT IS THE EXPECTED RESULT?
Decrypt data and get access to your phone and flash zips/roms...

#### WHAT HAPPENS INSTEAD?
Error "Can't decrypt data"

#### ADDITIONAL INFORMATION
ALatest June patch, latest TWRP version

I cant because i cant enter to twrp 

`/tmp/recovery.log`: 
`dmesg`:

**Comment by ash0x0** [comment]
Same here. Ran into it today with same info @TheGabrielHoward reported. I'd note though this probably didn't happen with 3.3.0 before the latest June update (PQ3A.190605.003), 3.3.0 ran fine on my previous October '18 PPR2.181005.003 build. I updated to June just today and installed 3.3.0 and it won't accept the correct pattern to decrypt, even though it works with the system.

EDIT: This same issue exists with all versions on the device download page. I tried all with pattern and password unlock as 3.1.1 doesn't seem to support pattern, problem exists with all. This probably has to do with recent changes in security updates between October and June. I'll try to flash October on other boot slot to confirm.

EDIT 2: Tried 3.3.0-0 with PPR2.181005.003, Oct 2018 and it decrypts fine. This probably has to do with the way twrp reads the encryption pass from the device, something was changed somewhere between October and June security updates that broke it.

**Comment by TheGabrielHoward** [comment]
> Same here. Ran into it today with same info @TheGabrielHoward reported. I'd note though this probably didn't happen with 3.3.0 before the latest June update (PQ3A.190605.003), 3.3.0 ran fine on my previous October '18 PPR2.181005.003 build. I updated to June just today and installed 3.3.0 and it won't accept the correct pattern to decrypt, even though it works with the system.
> 
> EDIT: This same issue exists with all versions on the device download page. I tried all with pattern and password unlock as 3.1.1 doesn't seem to support pattern, problem exists with all. This probably has to do with recent changes in security updates between October and June. I'll try to flash October on other boot slot to confirm.
> 
> EDIT 2: Tried 3.3.0-0 with PPR2.181005.003, Oct 2018 and it decrypts fine. This probably has to do with the way twrp reads the encryption pass from the device, something was changed somewhere between October and June security updates that broke it.

Yes, something changes. So while this is not fixed, to flash something I have to remove the security

**Comment by ash0x0** [comment]
@TheGabrielHoward How did you manage that? When I tried removing security my device remained encrypted and on the next reboot I was asked for the decryption pass even though they're no security after that. When I tried to remove the certificate and set device to unencrypted I was completely locked out on the next reboot. I have a verizon pixel so I attributed to some carrier certificate somewhere. What process did you follow to remove security?

**Comment by bigbiff** [comment]
Hi, can people who are having problems test http://build.twrp.me/twrp-3.3.1-enc-marlin.img

**Comment by ash0x0** [comment]
It's corrupted, can't open it.

**Comment by bigbiff** [comment]
Sorry, fixed link above

**Comment by ash0x0** [comment]
@bigbiff decryption works now and it reads storage like normal. I didn't try it further than this (no flashing, etc) because I didn't want to risk with a test build but as far as decryption goes you fixed it. Thank you so much for taking the time.

**Comment by TheGabrielHoward** [comment]
> @TheGabrielHoward How did you manage that? When I tried removing security my device remained encrypted and on the next reboot I was asked for the decryption pass even though they're no security after that. When I tried to remove the certificate and set device to unencrypted I was completely locked out on the next reboot. I have a verizon pixel so I attributed to some carrier certificate somewhere. What process did you follow to remove security?

i just went to security and set screen lock to none, then i went to the bootloader and select recovery mode (TWRP). TWRP does not ask me for password

**Comment by TheGabrielHoward** [comment]
> Hi, can people who are having problems test http://build.twrp.me/twrp-3.3.1-enc-marlin.img

its for fastboot? or twrp installer

**Comment by TheGabrielHoward** [comment]
> Hi, can people who are having problems test http://build.twrp.me/twrp-3.3.1-enc-marlin.img

yh, its working fine, it decrypts fine, now we need the installer. thanks

**Comment by bigbiff** [comment]
Thanks goes to Peter Cai: https://gerrit.omnirom.org/c/android_bootable_recovery/+/34092

**Comment by TheGabrielHoward** [comment]
@bigbiff Update: Tried to root with Magisk with encrypted phone using this recovery and i got bootloop

**Comment by ThauEx** [comment]
@bigbiff Could you please provide a fixed version for walleye(Pixel 2), too?

**Comment by TheArkive** [comment]
The latest bootloader I can use is PQ2A.190405.003, Apr 2019, May and June give me issues with decryption.  Using TWRP 3.3.0-0.  Hope to see Pixel fixed soon as well.

**Comment by mabz2k** [comment]
I have the same problem with LG Nexus 5 [HAMMERHEAD], using the latest recovery 3.3.1-0 and android 8.1 with the rom unlegacy. It wont decrypt data, have tried recovery with 3.2.2 and 3.1.1 and still the same problem. tried android 8.0 and still nothing. Is it ok to post this here since it sounds the same or would I need to make a new issue? new here so not sure. thank you in advance.

**Comment by clayissimo** [comment]
> Thanks goes to Peter Cai: https://gerrit.omnirom.org/c/android_bootable_recovery/+/34092

@bigbiff booting off "twrp-3.3.1-1-marlin.img" works fine including decrypting. However, the installer "twrp-installer-3.3.1-1-marlin.zip" (and "twrp-installer-3.3.1-0-marlin.zip") does not work! It does not boot properly. The last correctly working installer is "twrp-pixel-installer-marlin-3.3.0-0.zip", but that version does not decrypt correctly.
Any chance that you could have a look at that issue?

**Comment by TheArkive** [comment]
> 
> 
> > Thanks goes to Peter Cai: https://gerrit.omnirom.org/c/android_bootable_recovery/+/34092
> 
> @bigbiff booting off "twrp-3.3.1-1-marlin.img" works fine including decrypting. However, the installer "twrp-installer-3.3.1-1-marlin.zip" (and "twrp-installer-3.3.1-0-marlin.zip") does not work! It does not boot properly. The last correctly working installer is "twrp-pixel-installer-marlin-3.3.0-0.zip", but that version does not decrypt correctly.
> Any chance that you could have a look at that issue?

Which bootloader are you using that happens to work 3.3.1-1?

**Comment by clayissimo** [comment]
> > > Thanks goes to Peter Cai: https://gerrit.omnirom.org/c/android_bootable_recovery/+/34092
> > 
> > 
> > @bigbiff booting off "twrp-3.3.1-1-marlin.img" works fine including decrypting. However, the installer "twrp-installer-3.3.1-1-marlin.zip" (and "twrp-installer-3.3.1-0-marlin.zip") does not work! It does not boot properly. The last correctly working installer is "twrp-pixel-installer-marlin-3.3.0-0.zip", but that version does not decrypt correctly.
> > Any chance that you could have a look at that issue?
> 
> Which bootloader are you using that happens to work 3.3.1-1?

@TheArkive I am using the original bootloader from QP1A.190711.020 (bootloader-marlin-8996-012001-1907011432.img)

Just to make sure: When I say I can boot off "twrp-3.3.1-1-marlin.img" I mean booting in the sense of "C:\Unified_Android_ToolKit>.\fastboot-toolkit.exe boot twrp-3.3.1-1-marlin.img". 

Installing twrp-installer-3.3.1-1-marlin.zip and trying to boot into recovery TWRP does not work!

Does that make sense?

**Comment by TheArkive** [comment]
> 
> 
> > > > Thanks goes to Peter Cai: https://gerrit.omnirom.org/c/android_bootable_recovery/+/34092
> > > 
> > > 
> > > @bigbiff booting off "twrp-3.3.1-1-marlin.img" works fine including decrypting. However, the installer "twrp-installer-3.3.1-1-marlin.zip" (and "twrp-installer-3.3.1-0-marlin.zip") does not work! It does not boot properly. The last correctly working installer is "twrp-pixel-installer-marlin-3.3.0-0.zip", but that version does not decrypt correctly.
> > > Any chance that you could have a look at that issue?
> > 
> > 
> > Which bootloader are you using that happens to work 3.3.1-1?
> 
> @TheArkive I am using the original bootloader from QP1A.190711.020 (bootloader-marlin-8996-012001-1907011432.img)
> 
> Just to make sure: When I say I can boot off "twrp-3.3.1-1-marlin.img" I mean booting in the sense of "C:\Unified_Android_ToolKit>.\fastboot-toolkit.exe boot twrp-3.3.1-1-marlin.img".
> 
> Installing twrp-installer-3.3.1-1-marlin.zip and trying to boot into recovery TWRP does not work!
> 
> Does that make sense?

Yes indeed.  So decrypt is working with Android 10 already but not on Android 9?  funny...  I don't envy the devs ... must be like playing the most annoying game of wack-a-mole.

**Comment by clayissimo** [comment]
@TheArkive IIRC, it worked fine for me on Android 9.

**Comment by TheArkive** [comment]
You mean 3.3.1-1 worked fine in android 9?  Using latest bootloaders in Android 9 too?  Latest Android 9 bootloader I could use with 3.3.1-0 was April 2019 (PQ2A.190405.003).

**Comment by TheGabrielHoward** [comment]
> > Thanks goes to Peter Cai: https://gerrit.omnirom.org/c/android_bootable_recovery/+/34092
> 
> @bigbiff booting off "twrp-3.3.1-1-marlin.img" works fine including decrypting. However, the installer "twrp-installer-3.3.1-1-marlin.zip" (and "twrp-installer-3.3.1-0-marlin.zip") does not work! It does not boot properly. The last correctly working installer is "twrp-pixel-installer-marlin-3.3.0-0.zip", but that version does not decrypt correctly.
> Any chance that you could have a look at that issue?

That TWRP just decrypt, all files you install like twrp and magisk will not work and will corrupt the data

**Comment by clayissimo** [comment]
@TheGabrielHoward, not sure what you mean. Booting off 3.3.1.1 works fine, modifications are fine. Everything is OK. I just can't install 3.3.1.1 permanently properly.

**Comment by adl1995** [comment]
Just to add, TWRP decryption is not working on Pixel XL either. I tried using several versions (3.3.1-1-marlin, 3.2.2-0-marlin, 3.2.3-0-marlin, 3.2.3-0-marlin). I'm using the OTA from Google (version: 10.0.0 (QP1A.190711.020, Sep 2019)).

**Comment by clayissimo** [comment]
@adl1995 @TheGabrielHoward @TheArkive @bigbiff :
Sorry, something has been broken by trying to install 3.3.1.1. Since then I can't install magisk properly anymore. At the end of the installer I get an error message that /system and another partition can't be changed. 

Is TWRP broken on 10.0 with pixel xl? I am 💯 sure it worked before installing 3.3.1.1 because I used titanium backup to recover all my apps. Now it's not anymore.

**Comment by ash0x0** [comment]
I can reproduce the same state as @clayissimo with 3.3.1.1
Decryption and flashing work fine, permanent installer doesn't install. No errors or issues when installing though, says it installed successfully but didn't. Not a slot issue either, tried both slots in case it failed to flash to both. This seems to be another issue now so you can either close this issue, mark as solved and open another issue for the installer. That is if no one else is still reporting the decryption problem.

**Comment by DiThi** [comment]
Decryption is broken for sailfish with Android 10 (OTA update). It says "Pattern failed, please try again!".

Would it be enough to compile the latest changes from marlin into a build for sailfish? They seem to exist in the same repository. Latest build for sailfish is 3.3.0-0.

I use it only for booting with fastboot, no installation.
